src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -178,6 +178,32 @@ public DynamicObject fromRope(Rope rope) {
             return executeMake(rope, NotProvided.INSTANCE, NotProvided.INSTANCE);
         }
 
+        public DynamicObject fromBuilder(RopeBuilder builder, CodeRange codeRange) {
+            return executeMake(builder.getBytes(), builder.getEncoding(), codeRange);
+        }
+
+        /**
+         * All callers of this factory method must guarantee that the builder's byte array cannot change after this
+         * call, otherwise the rope built from the builder will end up in an inconsistent state.
+         */
+        public DynamicObject fromBuilderUnsafe(RopeBuilder builder, CodeRange codeRange) {
+            final byte[] unsafeBytes = builder.getUnsafeBytes();
+            final byte[] ropeBytes;
+
+            // While the caller must guarantee the builder's byte[] cannot change after this call, it's possible
+            // the builder has allocated more space than it needs. Ropes require that the backing byte array
+            // is the exact length required. If the builder doesn't satisfy this constraint, we must make a copy.
+            // Alternatively, we could make a leaf rope and then take a substring of it, but that would complicate
+            // the specializations here.
+            if (unsafeBytes.length == builder.getLength()) {
+                ropeBytes = unsafeBytes;
+            } else {
+                ropeBytes = builder.getBytes();
+            }
```

Comment:
Should this be profiled, for example by using a ConditionProfile in the node or using a wrapper node?

Reply:
That would be ideal, but every caller is behind a boundary right now. So I really only focused on eliminating the unnecessary `byte[]` copy. If you feel strongly about it, I can add separate specializations for the two branches.

Reply:
It seems fine then.
A good way is indeed to move that part of the logic in specializations by passing the builder down.